hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -978,12 +978,15 @@ public void createDirectory(final Path path, final FileSystem.Statistics statist
           throws IOException {
     try (AbfsPerfInfo perfInfo = startTracking("createDirectory", "createPath")) {
       if (getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB) {
-        checkParentChainForFile(path, tracingContext);
+        ArrayList<Path> keysToCreateAsFolder = new ArrayList<>();
+        checkParentChainForFile(path, tracingContext, keysToCreateAsFolder);
 
         HashMap<String, String> metadata = new HashMap<>();
         metadata.put(X_MS_META_HDI_ISFOLDER, TRUE);
-        createFile(path, statistics, true,
-                permission, umask, tracingContext, metadata);
+        for (Path pathToCreate: keysToCreateAsFolder) {
+          createFile(pathToCreate, statistics, false,
```

Comment:
Same fileAlreadyException can be here as well. 

We should throw fileAlreadyException to end dev only when the terminal dir is already there. For the parent dirs, we should just consume it.